### PR TITLE
League functionalities

### PR DESCRIPTION
**Pull request by kjprajapati18**
Ready to Merge!

TLDR:
- Created a basic User model and basic League model (minor changes may be needed)
- Finished league create, join, view (get list), leave, and disband. Only backend functions finished
- There are more functions/validations that will need to be created as our front-end and other features get added

Detailed:
- "/league/create" endpoint successfully creates a league, automatically assigning the creator to be both the manager and the first player
-  "/league/join" endpoint allows users to join league using leagueName and Key
- "/league/leave" endpoint allows users to leave league
- "/league/disband" endpoint allows league managers to completely delete a league
- "/league/find" endpoint allows us to pull a list of leagues for front-end display
- Create ensures that league names are unique
- Join/Leave/Disband ensures that leagues are properly deleted and that user's profile are updated
- League Management requires JWT Token for successful creation
- Updated validation.js to enforce authorization in request and prevent server from crashing (authValidation)
- Updated validation.js to ensure that league names are unique in the database (leagueCreation)
- Updated validation.js to ensure that usernames and emails are unique in the database (signup)
- All endpoints now properly return the correct success and error messages for each request

Known To-Do (later):
- [x] joinLeague should use leagueName to request to join. If the league is private, leagueKey is also required
- [x] Implement viewLeague with query parameters
- [x] Disband league option for league managers
- [x] Implement leaveLeague
- [x] Find Better way to validate Authorization token
- [ ] createLeague should add AI Players to league upon request (not yet implementable, waiting for AI implementation/discussion)

### CODE REVIEW

File: app/services/league/controllers/leagueController.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        { username: user },
+        { $addToSet: { leagues: leagueID } },
+        {},
+        (err) => {
+            if (err) throw err;
+        },
+    ),
+]);
+
+exports.authorizeLeague = async (req, res, next) => {
```
Comment (avidlaud) [inline app/services/league/controllers/leagueController.js:23]: Should this be in middleware instead of controller?

File: app/services/league/controllers/leagueController.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    }
+};
+
+exports.joinLeague = async (req, res) => {
+    const { _id } = res.locals.league;
+    const { username } = res.locals;
+
+    addPlayerToLeague(username, _id)
+        .then((succ) => {
+            if (!res.locals.league.playerList.includes(username)) {
```
Comment (avidlaud) [inline app/services/league/controllers/leagueController.js:65]: Maybe check before adding player to league

File: app/services/league/controllers/leagueController.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            { $pull: { leagues: _id } },
+            { new: true },
+            (err) => {
+                if (err) throw err;
+            },
+        ),
+    ]);
+    request
+        .then((succ) => {
+            if (res.locals.league.playerList.includes(username)) {
```
Comment (avidlaud) [inline app/services/league/controllers/leagueController.js:97]: Same here - check before leaving

File: app/services/league/controllers/leagueController.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        if (err) throw err;
+        res.send(result);
+    });
+};
+
+exports.disbandLeague = async (req, res) => {
+    const { _id, leagueManager: manager } = res.locals.league;
+    const { username } = res.locals;
+
+    if (manager !== username) {
```
Comment (avidlaud) [inline app/services/league/controllers/leagueController.js:123]: Not sure if we need to handle upper/lowercase - I think we should be good

File: app/services/league/models/leagueModel.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    },
+    leagueKey: {
+        type: String,
+        default: '',
+    },
+    playerList: {
+        type: Array,
+        default: [],
+    },
+    settings: [{
```
Comment (avidlaud) [inline app/services/league/models/leagueModel.js:21]: This should just be an object instead of an array

File: app/services/league/controllers/leagueController.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+exports.authorizeLeague = async (req, res, next) => {
+    const { _id: leagueID, leagueKey } = res.locals.league;
+
+    await League.findById(
+        leagueID,
+        (err) => {
+            if (err) throw err;
+        },
+    ).then((result) => {
+        if (result.settings[0].endDate.getTime() < Date.now()) return res.status(400).send('Error: You tried to join an already expired league');
```
Comment (avidlaud) [inline app/services/league/controllers/leagueController.js:32]: See comment in leaguemodel.js

File: frontend/src/components/Home/Home.js [hunk 7]
```diff
@@ -5,6 +5,9 @@ function Home() {
     if (!localStorage.getItem('token')) {
         return <Redirect to="/user/login" />;
     }
+    const token = localStorage.getItem('token');
+    console.log(token);
```
Comment (avidlaud) [inline frontend/src/components/Home/Home.js:9]: Comment out

File: app/services/league/controllers/leagueController.js (hunk 4 above)
Comment (kjprajapati18) [reply app/services/league/controllers/leagueController.js:123]: I think we're fine here as well

**Review by avidlaud — APPROVED** [review]

**Review by RiyaTayal20 — APPROVED** [review]

**Review by christinem61 — APPROVED** [review]
